File: WdtcCore/src/main/java/org/wdt/wdtc/core/launch/AbstractGameCommand.java
package org.wdt.wdtc.core.launch;

import org.wdt.wdtc.core.utils.StringUtils;

import java.io.File;
import java.io.IOException;

abstract class AbstractGameCommand {
  protected final StringBuilder command;

  public AbstractGameCommand() {
    command = new StringBuilder();
  }

  abstract StringBuilder getCommand() throws IOException;

  protected void nonBreakingSpace(Object o) {
    command.append(o).append(StringUtils.STRING_SPACE);
  }

  protected void insertclasspathSeparator(File file) {
    insertclasspathSeparator(file.getAbsolutePath());
  }

  protected void insertclasspathSeparator(String str) {
    command.append(str).append(";");
  }

  protected void insertSpace(String str) {
    command.append(str).append(StringUtils.STRING_SPACE);
  }

  protected void nonBreakingSpace(String str, Object string) {
    nonBreakingSpace(str + string);
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/launch/AbstractGameCommand.kt
package org.wdt.wdtc.core.launch

import org.wdt.wdtc.core.utils.STRING_SPACE

abstract class AbstractGameCommand {
  protected val commandBuilder: StringBuilder = StringBuilder()
  private val regex = ".*\\\$\\{(.*)}".toRegex()

  abstract fun getCommand(): StringBuilder

  abstract val dataMap: Map<String, String>
  protected fun nonBreakingSpace(o: Any) {
    commandBuilder.append(o).append(STRING_SPACE)
  }

  protected fun nonBreakingSpace(str: String, o: Any) {
    commandBuilder.append(str).append(o).append(STRING_SPACE)
  }

  protected fun String.replaceData(dataMap: Map<String, String>): String {
    if (this.matches(regex)) {
      return regex.find(this)?.groups?.get(1)?.value.let { str ->
        dataMap[str].let {
          if (it == null) this
          else this.replace("\${$str}", it)
        }
      }
    }
    return this
  }
}
